ls1intum/ArTEMiS
# Improve test cases for programming exercises

### Checklist

### Motivation and Context
This pull request improves the test cases for Java Programming Exercises.
It also provides the possibility to automatically generate the test.json after changing template and solution source code. This saves time for the instructors and provides better feedback to students

### Description
Update the structure and behavior tests in the template for Java Programming Exercises.
Provide a possibility to automatically generate the test.json oracle for structural test cases.

### Steps for Testing
1. Log in to ArTEMiS
2. Navigate to Course Administration
3. Create a Programming Exercise with Java (please use atest in the course short name and exercise short name)
4. Update the source code in the template and solution repositories (e.g. on Bitbucket or after cloning): introduce new classes, attributes, methods, enums etc. in the solution
5. Navigate into the programming exercise detail view (click on View in Artemis)
6. Click on the button 'Update Structure Test Oracle'
7. Check if the file test.json was updated according to the changes in 4.
8. Check if template and solution build plans are still compiling. The template build plan should have the result X of X tests failed (red, i.e. all tests should fail). The solution build plan should the result All tests passed (green)
9. Delete the programming exercise again (make sure to choose all ticks in the Delete dialog)
